### PR TITLE
Add error handling to GET requests

### PR DESCRIPTION
**Pull request by zurbergram**
Resolves https://vajira.max.gov/browse/APPEALS-22505, https://vajira.max.gov/browse/APPEALS-23001

# Description
- Add metrics database table
- Add logging endpoint to store data from frontend into metrics table
- Add error handling on all REST calls in ApiUtil.js

## Acceptance Criteria
- [ ] Code compiles correctly

## Testing Plan
1. Go to #{jira issue link} or list them below

- [ ] For feature branches merging into master: Was this deployed to UAT?

# Frontend
## User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 BEFORE|AFTER
 ---|---

## Storybook Story
*For Frontend (Presentationa) Components*
* [ ] Add a [Storybook](https://github.com/department-of-veterans-affairs/caseflow/wiki/Documenting-React-Components-with-Storybook) file alongside the component file (e.g. create `MyComponent.stories.js` alongside `MyComponent.jsx`)
* [ ] Give it a title that reflects the component's location within the overall Caseflow hierarchy
* [ ] Write a separate story (within the same file) for each discrete variation of the component

# Backend
## Database Changes
*Only for Schema Changes*

* [ ] Add typical timestamps (`created_at`, `updated_at`) for new tables
* [ ] Update column comments; include a "PII" prefix to indicate definite or potential [PII data content](https://github.com/department-of-veterans-affairs/appeals-team/blob/master/caseflow-team/0-how-we-work/pii-handbook.md#what-is-pii)
* [ ] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`) (see [Writing DB migrations](https://github.com/department-of-veterans-affairs/caseflow/wiki/Writing-DB-migrations))
* [ ] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [ ] Perform query profiling (eyeball Rails log, check bullet and fasterer output)
* [ ] For queries using raw sql was an explain plan run by System Team
* [ ] Add appropriate indexes (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [ ] Run `make check-fks`; add any missing foreign keys or add to `config/initializers/immigrant.rb` (see [Record associations and Foreign Keys](https://github.com/department-of-veterans-affairs/caseflow/wiki/Record-associations-and-Foreign-Keys))
* [ ] Add `belongs_to` for associations to enable the [schema diagrams](https://department-of-veterans-affairs.github.io/caseflow/task_trees/schema/schema_diagrams) to be automatically updated
* [ ] Document any non-obvious semantics or logic useful for interpreting database data at [Caseflow Data Model and Dictionary](https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Data-Model-and-Dictionary)

## Integrations: Adding endpoints for external APIs
* [ ] Check that Caseflow's external API code for the endpoint matches the code in the relevant integration repo
  * [ ] Request: Service name, method name, input field names
  * [ ] Response: Check expected data structure
  * [ ] Check that calls are wrapped in MetricService record block
* [ ] Check that all configuration is coming from ENV variables
  * [ ] Listed all new ENV variables in description
  * [ ] Worked with or notified System Team that new ENV variables need to be set
* [ ] Update Fakes
* [ ] For feature branches: Was this tested in Caseflow UAT

# Best practices
## Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [ ] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Monitoring
- [ ] Are performance metrics (e.g., response time, throughput) being tracked?
- [ ] Are key application components monitored (e.g., database, cache, queues)?
- [ ] Is there a system in place for setting up alerts based on performance thresholds?

### Logging
- [ ] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

File: client/app/2.0/components/reader/DocumentViewer/PDF/index.jsx [hunk 1]
```diff
@@ -67,7 +67,7 @@ export const Pdf = ({ doc, clickPage, ...props }) => {
               <StatusMessage title="Unable to load document" type="warning">
               Caseflow is experiencing technical difficulties and cannot load <strong>{doc.type}</strong>.
                 <br />
-              You can try <a href={`${doc.content_url}?type=${doc.type}&download=true`}>downloading the document</a>
+              You can try <a href={`${doc.content_url}?type=${doc.type}&download=true`}>downloading the document</a>{' '}
```
Comment (mikefinneran) [inline client/app/2.0/components/reader/DocumentViewer/PDF/index.jsx:70]: Whats the lil fella all about? Were we not getting a space between `document` and `or`?

File: db/schema.rb [hunk 2]
```diff
@@ -704,6 +704,7 @@
   end
 
   create_table "documents", id: :serial, force: :cascade do |t|
+    t.boolean "auto_tagged", default: false
```
Comment (mikefinneran) [inline db/schema.rb:707]: What is `auto_tagged` column used for?

File: db/schema.rb [hunk 3]
```diff
@@ -100,7 +100,7 @@
     t.boolean "privacy_act_complete", default: false, null: false, comment: "When true, appeal has a privacy act request completed"
     t.boolean "privacy_act_pending", default: false, null: false, comment: "When true, appeal has a privacy act request still open"
     t.boolean "scheduled_in_error", default: false, null: false, comment: "When true, hearing was scheduled in error and none scheduled"
-    t.datetime "updated_at", comment: "Date and time the record was last updated"
+    t.datetime "updated_at"
```
Comment (mikefinneran) [inline db/schema.rb:103]: Might have lost this comment during a migration but its also a self explanatory column so might be ok to lose.

File: client/app/util/Metrics.js [hunk 4]
```diff
@@ -31,3 +32,30 @@ export const collectHistogram = (data) => {
...
+  let id = uniqueId;
+
+  // If a uuid wasn't provided assume that metric also wasn't sent to javascript console and send with UUID to console
+  if (!uniqueId) {
+    id = uuid.v4();
+    if (isError) {
+      console.error(`${id}\n${data}`);
+    } else {
+      // eslint-disable-next-line no-console
+      console.log(`${id}\n${data}`);
```
Comment (amybids) [inline client/app/util/Metrics.js:46]: Why are we console logging this if it succeeds? Could we do a process.env lookup and not do the console.log in prod?

**Review by amybids — CHANGES_REQUESTED** [review]

File: client/app/2.0/components/reader/DocumentViewer/PDF/index.jsx (hunk 1 above)
Comment (zurbergram) [reply client/app/2.0/components/reader/DocumentViewer/PDF/index.jsx:70]: Correct. before screenshot
Screenshot 2023-05-25 at 4 26 19 PM

File: db/schema.rb (hunk 2 above)
Comment (zurbergram) [reply db/schema.rb:707]: old, not sure how it got in there

File: db/schema.rb (hunk 3 above)
Comment (zurbergram) [reply db/schema.rb:103]: I noticed this happens from time to time, comments are added or lost

File: client/app/util/Metrics.js (hunk 4 above)
Comment (zurbergram) [reply client/app/util/Metrics.js:46]: This is not used currently just building out something for future use. Goal is to dump metrics on console so can be viewed by Help Desk or Dev team

File: app/models/metric.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+class Metric < CaseflowRecord
+  belongs_to :user
+
+  METRIC_TYPES = { error: 'error', log: 'log', performance: 'performance' }
+  LOG_SYSTEMS = { datadog: 'datadog', rails_console: 'rails_console', javascript_console: 'javascript_console' }
```
[the remark this answers is not in the document]
Comment (zurbergram) [reply app/models/metric.rb]: changed

**Pull request merged**